src/main/java/org/springframework/retry/annotation/AnnotationAwareRetryOperationsInterceptor.java

```diff
@@ -416,4 +422,27 @@ private String resolve(String value) {
 		return value;
 	}
 
+	/**
+	 * Check if the expression is a template
+	 * @param expression the expression string
+	 * @return true if the expression string is a template
+	 */
+	private boolean isTemplate(String expression) {
+		return expression.contains(PARSER_CONTEXT.getExpressionPrefix()) && expression.contains(PARSER_CONTEXT.getExpressionSuffix());
+	}
+
+	/**
+	 * Get expression based on the expression string.
+	 * At the moment supports both literal and template expressions. Template expressions are deprecated.
+	 * @param expression the expression string
+	 * @return literal expression or template expression
+	 */
+	private Expression getExpression(String expression) {
+		if (isTemplate(expression)) {
+			logger.warn("Template expressions are deprecated in favor of literal expressions");
```

Comment:
I wouldn't say this here.
As I said before, unlike `ExpressionRetryPolicy` all the other expressions are evaluated only once at start up when the annotation is parsed.

In other projects we follow a convention like support SpEL and property placeholder in the main property suffering from the fact that it can be only as a string.
For example `@RabbitListener`:
```
	/**
	 * Set to true or false, to override the default setting in the container factory.
	 * @return true to auto start, false to not auto start.
	 * @since 2.0
	 */
	String autoStartup() default "";
```
and in the `RabbitListenerAnnotationBeanPostProcessor` we do this:
```
String autoStartup = rabbitListener.autoStartup();
if (StringUtils.hasText(autoStartup)) {
	endpoint.setAutoStartup(resolveExpressionAsBoolean(autoStartup));
}
...
private Object resolveExpression(String value) {
	String resolvedValue = resolve(value);

	return this.resolver.evaluate(resolvedValue, this.expressionContext);
}

private String resolve(String value) {
	if (this.beanFactory != null && this.beanFactory instanceof ConfigurableBeanFactory) {
		return ((ConfigurableBeanFactory) this.beanFactory).resolveEmbeddedValue(value);
	}
	return value;
}
```

In other words we don't need to have here this in the `@Retryable`:
```
int maxAttempts() default 3;

String maxAttemptsExpression() default "";
```

The single `String maxAttempts() default "3";` may cover all the use-cases with simple value and expression. However again: this expression evaluated only one at start up. I call them *bean factory expresisons* since they are called against `BeanExpressionContext` and in the `StandardBeanExpressionResolver`.

I may agree that this my suggestion is going to be as a big breaking change so we may just stay with the current architecture until the next major version or  just forever...

Therefore my suggestion is to stay with templating here in the `AnnotationAwareRetryOperationsInterceptor` meanwhile supporting plain expressions. I mean there is nothing to deprecate in this class!

Reply:
If you look at the git blame, you will see that I added the `...Expression` flavors exactly because of the breaking change.

See https://github.com/spring-projects/spring-retry/issues/34

>I went with discrete attributes to avoid breaking changes.

I don't think they can be removed until, perhaps 2.0.

But I agree; since these expressions are evaluated once-only, they should remain as-is to be consistent with normal Spring configuration conventions.

Reply:
Great! Thank you Gary!

So, @aldex32 , according my concern and Gary's feedback, please, reinstate changes in the `AnnotationAwareRetryOperationsInterceptor`.
We will remove this expression variant in the future according my explanation in the previous comment. And in that case the target expressions will remain as templated. No other choice.

Thanks for understanding and sorry for confusing if that.

Reply:
done